### DISCUSSION TITLE
Incompatibility between kivy garden matplotlib and pyinstaller

### DISCUSSION THREAD

**Discussion by Patricklevyy** [pyinstaller/pyinstaller#8078] [Help] [answered]
Hello, is there any incompatibility between a kivy garden matplotlib and pyinstaller? I have a kivy application that uses the module kivy.garden.matplotlib.backend_kivyagg. When I test my program it works fine but once converted into an executable (.exe) I get this error below. I'm using a virtual environment and I've checked that the module is correctly installed. I also run the pyinstaller command in the same viral environment.
```
Traceback (most recent call last):
  File "main.py", line 6, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "myapp\load_data_app.py", line 11, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "myapp\curve_drawing.py", line 5, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1140, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'kivy.garden.matplotlib.backend_kivyagg'
```

I use:

```
Kivy==2.2.1
Kivy-Garden==0.1.5
kivy-garden.graph==0.4.0
kivy-matplotlib-widget==0.6.1
kivymd==1.1.1
matplotlib==3.7.2
pyinstaller==5.6.2
```
Thank you for your help.

**Comment by rokm** [comment]
What does line 5 in `myapp\curve_drawing.py` look like?

**Reply by Patricklevyy** [reply]
It is `from kivy.garden.matplotlib.backend_kivyagg import FigureCanvasKivyAgg`

**Reply by rokm** [reply]
Hmmm, in that case, I don't suppose adding `kivy.garden.matplotlib.backend_kivyagg` to hiddenimports (e.g., adding `--hiddenimport kivy.garden.matplotlib.backend_kivyagg` to your PyInstaller command) makes any difference?

**Reply by Patricklevyy** [reply]
Yes it makes no difference

**Reply by rokm** [reply]
In that case, can you provide a complete minimal example (i.e., whole Kivy app that attempts to use `kivy.garden.matplotlib`)?

**Reply by Patricklevyy** [reply]
Here's an example:
```
from kivy.lang import Builder
from kivymd.app import MDApp
from kivy.uix.floatlayout import FloatLayout
import matplotlib.pyplot as plt
from kivy.garden.matplotlib.backend_kivyagg import FigureCanvasKivyAgg

KV = """
<Graph>:

    BoxLayout:
        id: box
        size_hint_y: .8
        pos_hint: {"top": 1}
    BoxLayout:
        size_hint_y: .2
        TextInput:
            id: namer
        Button:
            text:"Save it!"
            on_release: root.save_it()
"""

x = [1, 2, 3, 4, 5]
y = [5, 12, 6, 9, 15]

plt.plot(x, y)
plt.ylabel("Y")
plt.xlabel("X")

class Graph(FloatLayout):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)

        box = self.ids.box
        box.add_widget(FigureCanvasKivyAgg(plt.gcf()))

    def save_it(self):
        name = self.ids.namer.text
        if name:
            plt.savefig(name)

class MainApp(MDApp):
    def build(self):
        self.theme_cls.theme_style = "Dark"
        self.theme_cls.primary_palette = "BlueGray"
        Builder.load_string(KV)
        return Graph()
    
MainApp().run()
```

**Reply by rokm** [reply]
Hmm, I don't think this is going to work.

Assuming `kivy.garden.matplotlib` needs to be installed via `garden`, that places the modules into `~/.kivy/garden/garden.matplotlib` and they are turned into `kivy.garden.matplotlib` at run-time via custom `kivy.garden.GardenImporter`. So PyInstaller's analysis (which is still based on the old paths & Finder approach) will not find them - and even if it did, what would be the proper place to collect them in the frozen application bundle...?

So unless `kivy` (which provides its own PyInstaller hook) has explicit support for `kivy.garden` modules/packages, you're better off finding another tool for building your executable.

**Reply by rokm** [reply]
It seems `kivy.garden` has run-time support for PyInstaller:
https://github.com/kivy/kivy/blob/533a95fafaba167218dc614be0dfc240af7b54fd/kivy/garden/__init__.py#L152-L153

So the `garden` modules can be placed into `libs/garden` directory in the frozen application's top-level directory, at least manually (or via `--add-data`). That will bypass their dependency scanning, though, so you will need to deal with all missing modules via hidden imports.

**Reply by Patricklevyy** [reply]
Ok i see so like that?
`pyinstaller.exe -w --debug=all --add-data '.\env\Lib\site-packages\kivy\garden\matplotlib\backend_kivyagg.py;kivy/garden/matplotlib' --hidden-import 'kivy.garden.matplotlib.backend_kivyagg' main.py`

**Reply by rokm** [reply]
You probably need to collect the whole package, and the target directory needs to be `libs/garden`.:
`--add-data '.\env\Lib\site-packages\kivy\garden\matplotlib;libs\garden\matplotlib`
or even
`--add-data '.\env\Lib\site-packages\kivy\garden;libs\garden`

**Reply by rokm** [reply]
Although, if you have `matplotlib` in `site-packages\kivy\garden\matplotlib`, then `from kivy.garden.matplotlib import ...` *should* cause the module to be collected (and handled by PyInstaller's PyiFrozenImporter at run-time instead of kivy's GardenImporter)...

**Reply by Patricklevyy** [reply]
I tried `pyinstaller.exe -w --debug=all --add-data '.\env\Lib\site-packages\kivy\garden;libs\garden'  --hidden-import 'kivy.garden.matplotlib.backend_kivyagg' main.py` and `pyinstaller.exe -w --debug=all --add-data '.\env\Lib\site-packages\kivy\garden\matplotlib;libs\garden\matplotlib' --hidden-import 'kivy.garden.matplotlib.backend_kivyagg' main.py` but I still get the same error message

**Reply by Patricklevyy** [reply]
And I have  `matplotlib` in `site-packages\kivy\garden\matplotlib`

**Reply by rokm** [reply]
What are then contents of the following directories, then?
- `dist\<name>\kivy`
- `dist\<name>\kivy\garden`
- `dist\<name>\kivy\garden\matplotlib`
- `dist\<name>\libs\garden\matplotlib`

**Reply by rokm** [reply]
> And I have `matplotlib` in `site-packages\kivy\garden\matplotlib`

How did you install it, by the way?

**Reply by Patricklevyy** [reply]
In `dist\<name>\kivy` :
```
animation.pyc
app.pyc
atlas.pyc
base.pyc
cache.pyc
clock.pyc
compat.pyc
config.pyc
context.pyc
core
deps
effects
event.pyc
eventmanager
extras
factory.pyc
factory_registers.pyc
garden
gesture.pyc
graphics
input
lang
lib
loader.pyc
logger.pyc
metrics.pyc
modules
multistroke.pyc
parser.pyc
properties.cp311-win_amd64.pyd
resources.pyc
setupconfig.pyc
support.pyc
uix
utils.pyc
vector.pyc
weakmethod.pyc
weakproxy.cp311-win_amd64.pyd
_clock.cp311-win_amd64.pyd
_event.cp311-win_amd64.pyd
_metrics.cp311-win_amd64.pyd
_version.pyc
__init__.pyc
```

In `dist\<name>\kivy\garden` :
```
matplotlib
__init__.pyc
```

In `dist\<name>\kivy\garden\matplotlib` :
```
backend_kivy.pyc
backend_kivyagg.pyc
__init__.pyc

```
And i installed using `garden install matplotlib`

**Reply by rokm** [reply]
Looks like it is `garden install --kivy matplotlib` that installs it into `site-packages/kivy/garden`.

Anyway, I've now tried on both Linux and Windows (previously, I was trying only linux), and in both cases, plain `pyinstaller --clean --noconfirm program.py` builds the example you provided, and the executable works (the plot is displayed). 

So, can you try building the earlier example instead of your app - does that work in your case as well?

**Reply by Patricklevyy** [reply]
I think when i started looking for a solution i also did `garden install --kivy matplotlib`. I tried the command (`pyinstaller --clean --noconfirm main.py`) when I run the .exe file it closes automatically

**Reply by rokm** [reply]
If you don't use `--noconsole` / `--windowed` / `-W`, you need to run the exe from command prompt (`cmd`); otherwise you won't see uhnandled exceptions that terminate your app.

**Reply by Patricklevyy** [reply]
Okay i tried again with `pyinstaller --clean --noconfirm -w --debug=all main.py` and it's the same error. And the line 4 is `from kivy.garden.matplotlib.backend_kivyagg import FigureCanvasKivyAgg
`
```
Traceback (most recent call last):
  File "main.py", line 4, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1140, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'kivy.garden.matplotlib.backend_kivyagg'

```

**Reply by rokm** [reply]
I'm out of ideas, then. It works for me.

Can you archive both the `build` and `dist` directories, and upload the archive somewhere so I can take a look at the files?

**Reply by Patricklevyy** [reply]
https://drive.google.com/file/d/1tHBMKgCHOI4e3SL4x3eQRbuhLo-UCfTW/view?usp=drive_link
https://drive.google.com/file/d/19mIyXE1SPCKeR-lcSVOZoSw_nlM4Zkyj/view?usp=drive_link
Thank you for your help!

**Reply by rokm** [reply]
You need to grant me access on those...

**Reply by Patricklevyy** [reply]
I did it

**Reply by rokm** [reply]
Can you also do a clean build *without* `--debug=all` and without `-w`, and upload `build` and `dist` for that?

**Reply by Patricklevyy** [reply]
https://drive.google.com/file/d/1gDG6fYJGNE7eD_I-QykrZUmCbWdlsgru/view?usp=drive_link

**Reply by rokm** [reply, marked as answer]
Aha, so your use of `--debug=all` (which implies `--debug noarchive`) is actually masking another issue:

```
>main
[5936] Module object for pyimod02_importers is NULL!
Traceback (most recent call last):
  File "PyInstaller\loader\pyimod02_importers.py", line 22, in <module>
  File "pathlib.py", line 14, in <module>
  File "urllib\parse.py", line 40, in <module>
ModuleNotFoundError: No module named 'ipaddress'
Traceback (most recent call last):
  File "PyInstaller\loader\pyiboot01_bootstrap.py", line 17, in <module>
ModuleNotFoundError: No module named 'pyimod02_importers'
[5936] Failed to execute script 'pyiboot01_bootstrap' due to unhandled exception!
```

which is caused by you using an old version of PyInstaller (5.6.2) with python >= 3.11.4 (#7692). Can you try updating PyInstaller to at least 5.12 *or* modifying your local copy of 5.6.2 to add `'ipaddress'` to `PY3_BASE_MODULES` list in `PyInstaller.compat` ([here](https://github.com/pyinstaller/pyinstaller/blob/09b8a1ebd0a62c4e61de61cd33c739c997249a89/PyInstaller/compat.py#L617-L652)).

Then rebuild, again without `--debug=all` (which you should never use, really), and for now without `-w` as well. Then run the executable from the `cmd`. If it still doesn't work, upload the `build` and `dist` directories again.

**Reply by Patricklevyy** [reply]
THANK YOU SO MUCHHHHH for helping me! It works
